README.md
# Description

This is an item tracker for Metroid Fusion to use with [EmoTracker](https://emotracker.net). <s>It can be hooked with Bizhawk to provide auto item tracking.</s>

## Thanks

ChaosMiles07, Ridleymaster for the original SM icons (https://www.spriters-resource.com/custom_edited/metroidcustoms/sheet/23198/)


changelog.txt
1.2.0
Added settings menu (split security, require main missile, require main PB)
Now both layout are merged into one (use the settings menu to change the layout)
Moved all images into different folders

---------------
1.1.1
Made broadcast modifiable (for streamers)

---------------
1.1.0
Added auto tracking support

---------------
1.0.3
Added bosses icons, missile data and PB data icons
Replaced old icons with Jakoliath's icons

---------------
1.0.2
Added Split Lock Levels variant
Added black outline to Lock Level images

---------------
1.0.1
Added Lock Level icons

---------------
1.0.0
Initial Release

scripts/inc/memory.lua
function get_address(memory_type, address)
	if memory_type == "IRAM" then
		return 0x03000000 + address
	end
	if memory_type == "ROM" then
		return 0x08000000 + address
	end
	return nil
end

function read_u8(memory_type, address)
	return AutoTracker:ReadU8(get_address(memory_type, address), 0)
end

function read_u16(memory_type, address)
	return read_u8(memory_type, address + 1) << 8 |
		   read_u8(memory_type, address)
end

function read_u32(memory_type, address)
	return read_u8(memory_type, address + 3) << 24 |
           read_u8(memory_type, address + 2) << 16 |
		   read_u8(memory_type, address + 1) << 8 |
		   read_u8(memory_type, address)
end

function read_s8(memory_type, address)
	return AutoTracker:ReadS8(get_address(memory_type, address), 0)
end

function read_s16(memory_type, address)
	return read_s8(memory_type, address + 1) << 8 |
		   read_s8(memory_type, address)
end

function read_s32(memory_type, address)
	return read_s8(memory_type, address + 3) << 24 |
           read_s8(memory_type, address + 2) << 16 |
		   read_s8(memory_type, address + 1) << 8 |
		   read_s8(memory_type, address)
end

function read_char(memory_type, address)
	return string.char(read_u8(memory_type, address))
end

function read_string(memory_type, address, length)
	str = ""
	for i=1,length do
		str = str..read_char(memory_type, address+(i - 1))
	end
	return str
end

scripts/autotracking.lua
lock_level_icons = {}
default_tracker_height = 0
default_broadcast_tracker_height = 0

-- Invoked when the auto-tracker is activated/connected
function autotracker_started()
	initMemoryWatches()
end

--Invoked when the auto-tracker is stopped.
function autotracker_stopped()
	clearMemoryWatches()
end

function updateLockLevelIcons()
	if next(lock_level_icons) == nil then
		table.insert(lock_level_icons, {
			Tracker:FindObjectForCode("lock_level").Stages[1].Icon,
			Tracker:FindObjectForCode("lock_level").Stages[2].Icon,
			Tracker:FindObjectForCode("lock_level").Stages[3].Icon,
			Tracker:FindObjectForCode("lock_level").Stages[4].Icon
		})
		table.insert(lock_level_icons, {
			Tracker:FindObjectForCode("l1").Icon,
			Tracker:FindObjectForCode("l1").ActiveIcon
		})
		table.insert(lock_level_icons, {
			Tracker:FindObjectForCode("l2").Icon,
			Tracker:FindObjectForCode("l2").ActiveIcon
		})
		table.insert(lock_level_icons, {
			Tracker:FindObjectForCode("l3").Icon,
			Tracker:FindObjectForCode("l3").ActiveIcon
		})
		table.insert(lock_level_icons, {
			Tracker:FindObjectForCode("l4").Icon,
			Tracker:FindObjectForCode("l4").ActiveIcon
		})
		default_tracker_height = Layout:FindLayout("tracker_default").Root.Height
		default_broadcast_tracker_height = Layout:FindLayout("tracker_broadcast").Root.Height
	end
	if Tracker:FindObjectForCode("setting_split_security").CurrentStage == 1 then
		Tracker:FindObjectForCode("lock_level").Icon = nil
		Tracker:FindObjectForCode("lock_level").DisabledImageFilterSpec = nil
		Tracker:FindObjectForCode("lock_level").IgnoreUserInput = true
		Tracker:FindObjectForCode("l1").Icon = lock_level_icons[2][1]
		Tracker:FindObjectForCode("l1").ActiveIcon = lock_level_icons[2][2]
		Tracker:FindObjectForCode("l1").IgnoreUserInput = false
		Tracker:FindObjectForCode("l2").Icon = lock_level_icons[3][1]
		Tracker:FindObjectForCode("l2").ActiveIcon = lock_level_icons[3][2]
		Tracker:FindObjectForCode("l2").IgnoreUserInput = false
		Tracker:FindObjectForCode("l3").Icon = lock_level_icons[4][1]
		Tracker:FindObjectForCode("l3").ActiveIcon = lock_level_icons[4][2]
		Tracker:FindObjectForCode("l3").IgnoreUserInput = false
		Tracker:FindObjectForCode("l4").Icon = lock_level_icons[5][1]
		Tracker:FindObjectForCode("l4").ActiveIcon = lock_level_icons[5][2]
		Tracker:FindObjectForCode("l4").IgnoreUserInput = false
		Layout:FindLayout("tracker_default").Root.Height = default_tracker_height
		Layout:FindLayout("tracker_broadcast").Root.Height = default_broadcast_tracker_height
	else
		if Tracker:FindObjectForCode("lock_level").Icon == nil then
			Tracker:FindObjectForCode("lock_level").CurrentStage = 1
			Tracker:FindObjectForCode("lock_level").CurrentStage = 0
		end
		Tracker:FindObjectForCode("lock_level").IgnoreUserInput = false
		Tracker:FindObjectForCode("l1").Icon = nil
		Tracker:FindObjectForCode("l1").IgnoreUserInput = true
		Tracker:FindObjectForCode("l2").Icon = nil
		Tracker:FindObjectForCode("l2").IgnoreUserInput = true
		Tracker:FindObjectForCode("l3").Icon = nil
		Tracker:FindObjectForCode("l3").IgnoreUserInput = true
		Tracker:FindObjectForCode("l4").Icon = nil
		Tracker:FindObjectForCode("l4").IgnoreUserInput = true
		Layout:FindLayout("tracker_default").Root.Height = default_tracker_height - 50
		Layout:FindLayout("tracker_broadcast").Root.Height = default_broadcast_tracker_height - 65
	end
end

--Invoked when clicking on the settings
function tracker_on_accessibility_updated()
    updateLockLevelIcons()
end

-- This is kind of hacky, but the tracker needs at least one memory watch set up in order to detect that the package
-- uses auto-tracking.  The only one that is active all the time is the "Which Game Is It Anyways" watch, but until we
-- get proper native addressing we can't initialize it until we know what connector is selected.
-- This is a dummy entry that will do nothing, the actual watches will be set when it starts above.
-- Set this watch interval to 60 seconds so it fires infrequently.
function noop()
    return true
end

ScriptHost:AddMemoryWatch("Let EmoTracker know we're here :)", 0x03000000, 0x02, noop, 60 * 1000)

scripts/inc/game.lua
missile_bomb_status = nil
max_missiles = nil
max_powerbombs = nil
beams_status = nil
suit_status = nil
max_etanks = nil
bosses_status = nil
lock_level = nil

-- Get the game identifier (aka the first 3 letters)
function getGameIdentifier()
	return read_string("ROM", OFF_ROM_GAMEIDENTIFIER, 3)
end

-- Get the game region (aka the fourth letter)
function getGameRegion()
	return read_char("ROM", OFF_ROM_GAMEIDENTIFIER + 3)
end

-- Get In Game Time as int32
function getIGT()
	return read_u8(OFF_IRAM_PLAYTIME_H) * H + read_u8(OFF_IRAM_PLAYTIME_M) * M + read_u8(OFF_IRAM_PLAYTIME_S) * S + (read_u8(OFF_IRAM_PLAYTIME_CS) * S) / 63
end

-- Is ingame
function isInGame()
	if getGameIdentifier() ~= "AMT" then
		return false
	end
	if getIGT() <= 16 then
		return false
	end
	GameMode = read_u16("IRAM", OFF_IRAM_GAME_MODE)
	return GameMode == 1 or GameMode == 3 or GameMode == 4 or GameMode == 5
end

function updateMissileBallStatus()
	local missileBallStatus = read_u8("IRAM", OFF_IRAM_MISSILES_BALL_STATUS)
	if isInGame() then
		-- Missile Launcher
		Tracker:FindObjectForCode("datam").Active = missileBallStatus & 1 == 1
		Tracker:FindObjectForCode("missile").AcquiredCount = read_u16("IRAM", OFF_IRAM_MAX_MISSILES)
		if Tracker:FindObjectForCode("setting_require_main_missile").CurrentStage then
			Tracker:FindObjectForCode("missile_icon").Active = missileBallStatus & 1 == 1
		else
			Tracker:FindObjectForCode("missile_icon").Active = Tracker:FindObjectForCode("missile").AcquiredCount > 0
		end
		-- Super Missile
		Tracker:FindObjectForCode("super_missile").Active = missileBallStatus & 2 == 2
		-- Ice Missile
		Tracker:FindObjectForCode("ice_missile").Active = missileBallStatus & 4 == 4
		-- Diffusion Missile
		Tracker:FindObjectForCode("diffusion_missile").Active = missileBallStatus & 8 == 8
		-- Morph Ball Bombs
		Tracker:FindObjectForCode("morph_ball_bombs").Active = missileBallStatus & 16 == 16
		-- Power Bombs
		Tracker:FindObjectForCode("datapb").Active = missileBallStatus & 32 == 32
		Tracker:FindObjectForCode("powerbomb").AcquiredCount = read_u8("IRAM", OFF_IRAM_MAX_POWERBOMBS)
		if Tracker:FindObjectForCode("setting_require_main_pb").CurrentStage then
			Tracker:FindObjectForCode("pb_icon").Active = missileBallStatus & 32 == 32
		else
			Tracker:FindObjectForCode("pb_icon").Active = Tracker:FindObjectForCode("powerbomb").AcquiredCount > 0
		end
	else
		-- Missile Launcher
		Tracker:FindObjectForCode("datam").Active = false
		Tracker:FindObjectForCode("missile_icon").Active = false
		Tracker:FindObjectForCode("missile").AcquiredCount = 0
		-- Super Missile
		Tracker:FindObjectForCode("super_missile").Active = false
		-- Ice Missile
		Tracker:FindObjectForCode("ice_missile").Active = false
		-- Diffusion Missile
		Tracker:FindObjectForCode("diffusion_missile").Active = false
		-- Morph Ball Bombs
		Tracker:FindObjectForCode("morph_ball_bombs").Active = false
		-- Power Bombs
		Tracker:FindObjectForCode("datapb").Active = false
		Tracker:FindObjectForCode("pb_icon").Active = false
		Tracker:FindObjectForCode("powerbomb").AcquiredCount = 0
	end
end

-- Get amount of energy tanks
function updateEtanks()
	local maxEtanks = (read_u16("IRAM", OFF_IRAM_MAX_HEALTH) - 99) / 100
	if isInGame() then
		Tracker:FindObjectForCode("etank").AcquiredCount = maxEtanks
		Tracker:FindObjectForCode("etank_icon").Active = maxEtanks > 0
	else
		Tracker:FindObjectForCode("etank").AcquiredCount = 0
		Tracker:FindObjectForCode("etank_icon").Active = false
	end
end

-- Get beam status
function updateBeamStatus()
	local beamStatus = read_u8("IRAM", OFF_IRAM_BEAMS_STATUS)
	if isInGame() then
		Tracker:FindObjectForCode("charge_beam").Active = beamStatus & 1 == 1
		Tracker:FindObjectForCode("wide_beam").Active = beamStatus & 2 == 2
		Tracker:FindObjectForCode("plasma_beam").Active = beamStatus & 4 == 4
		Tracker:FindObjectForCode("wave_beam").Active = beamStatus & 8 == 8
		Tracker:FindObjectForCode("ice_beam").Active = beamStatus & 16 == 16
	else
		Tracker:FindObjectForCode("charge_beam").Active = false
		Tracker:FindObjectForCode("wide_beam").Active = false
		Tracker:FindObjectForCode("plasma_beam").Active = false
		Tracker:FindObjectForCode("wave_beam").Active = false
		Tracker:FindObjectForCode("ice_beam").Active = false
	end
end

-- Get suit status
function updateSuitStatus()
	local suitStatus = read_u8("IRAM", OFF_IRAM_SUIT_STATUS)
	if isInGame() then
		Tracker:FindObjectForCode("hijump_boots").Active = suitStatus & 1 == 1
		Tracker:FindObjectForCode("speed_booster").Active = suitStatus & 2 == 2
		Tracker:FindObjectForCode("space_jump").Active = suitStatus & 4 == 4
		Tracker:FindObjectForCode("screw_attack").Active = suitStatus & 8 == 8
		Tracker:FindObjectForCode("varia_suit").Active = suitStatus & 16 == 16
		Tracker:FindObjectForCode("gravity_suit").Active = suitStatus & 32 == 32
		Tracker:FindObjectForCode("morph_ball").Active = suitStatus & 64 == 64
	else
		Tracker:FindObjectForCode("hijump_boots").Active = false
		Tracker:FindObjectForCode("speed_booster").Active = false
		Tracker:FindObjectForCode("space_jump").Active = false
		Tracker:FindObjectForCode("screw_attack").Active = false
		Tracker:FindObjectForCode("varia_suit").Active = false
		Tracker:FindObjectForCode("gravity_suit").Active = false
		Tracker:FindObjectForCode("morph_ball").Active = false
	end
end

-- Get boss status
function updateBossStatus()
	local bossStatus = read_u16("IRAM", OFF_IRAM_BOSS_STATUS)
	if isInGame() then
		Tracker:FindObjectForCode("arachnus").Active = bossStatus & 1 == 1
		Tracker:FindObjectForCode("torizo").Active = bossStatus & 2 == 2
		Tracker:FindObjectForCode("science").Active = bossStatus & 4 == 4
		Tracker:FindObjectForCode("nettori").Active = bossStatus & 8 == 8
		Tracker:FindObjectForCode("box2").Active = bossStatus & 16 == 16
		Tracker:FindObjectForCode("zazabi").Active = bossStatus & 32 == 32
		Tracker:FindObjectForCode("serris").Active = bossStatus & 64 == 64
		Tracker:FindObjectForCode("chungus").Active = bossStatus & 128 == 128
		Tracker:FindObjectForCode("yakuza").Active = bossStatus & 256 == 256
		Tracker:FindObjectForCode("nightmare").Active = bossStatus & 512 == 512
		Tracker:FindObjectForCode("ridley").Active = bossStatus & 1024 == 1024
		--Tracker:FindObjectForCode("box1").Active = bossStatus & 2048 == 2048
	else
		Tracker:FindObjectForCode("arachnus").Active = false
		Tracker:FindObjectForCode("torizo").Active = false
		Tracker:FindObjectForCode("science").Active = false
		Tracker:FindObjectForCode("nettori").Active = false
		Tracker:FindObjectForCode("box2").Active = false
		Tracker:FindObjectForCode("zazabi").Active = false
		Tracker:FindObjectForCode("serris").Active = false
		Tracker:FindObjectForCode("chungus").Active = false
		Tracker:FindObjectForCode("yakuza").Active = false
		Tracker:FindObjectForCode("nightmare").Active = false
		Tracker:FindObjectForCode("ridley").Active = false
		--Tracker:FindObjectForCode("box1").Active = false
	end
end

-- Get lock level
function updateLockLevel()
	local lockLevel = read_u8("IRAM", OFF_IRAM_LOCK_LEVEL)
	if Tracker:FindObjectForCode("setting_split_security").CurrentStage == 1 then
		if isInGame() then
			Tracker:FindObjectForCode("l1").Active = lockLevel & 2 == 2
			Tracker:FindObjectForCode("l2").Active = lockLevel & 4 == 4
			Tracker:FindObjectForCode("l3").Active = lockLevel & 8 == 8
			Tracker:FindObjectForCode("l4").Active = lockLevel & 16 == 16
		else
			Tracker:FindObjectForCode("l1").Active = false
			Tracker:FindObjectForCode("l2").Active = false
			Tracker:FindObjectForCode("l3").Active = false
			Tracker:FindObjectForCode("l4").Active = false
		end
		Tracker:FindObjectForCode("lock_level").CurrentStage = 0
	else
		-- To prevent lock level to be -1
		if lockLevel < 0 then
			lockLevel = 0
		end
		if isInGame() then
			Tracker:FindObjectForCode("lock_level").CurrentStage = lockLevel
		else
			Tracker:FindObjectForCode("lock_level").CurrentStage = 0
		end
		Tracker:FindObjectForCode("l1").Active = false
		Tracker:FindObjectForCode("l2").Active = false
		Tracker:FindObjectForCode("l3").Active = false
		Tracker:FindObjectForCode("l4").Active = false
	end
end

-- Initialize memory watches
function initMemoryWatches()
	missile_bomb_status = ScriptHost:AddMemoryWatch("Missile/Bomb status", get_address("IRAM", OFF_IRAM_MISSILES_BALL_STATUS), 1, updateMissileBallStatus, 60)
	max_missiles = ScriptHost:AddMemoryWatch("Max Missiles", get_address("IRAM", OFF_IRAM_MAX_MISSILES), 2, updateMissileBallStatus, 60)
	max_powerbombs = ScriptHost:AddMemoryWatch("Max Power Bombs", get_address("IRAM", OFF_IRAM_MAX_POWERBOMBS), 1, updateMissileBallStatus, 60)
	max_etanks = ScriptHost:AddMemoryWatch("Max Etanks", get_address("IRAM", OFF_IRAM_MAX_HEALTH), 2, updateEtanks, 60)
	beam_status = ScriptHost:AddMemoryWatch("Beam status", get_address("IRAM", OFF_IRAM_BEAMS_STATUS), 1, updateBeamStatus, 60)
	suit_status = ScriptHost:AddMemoryWatch("Suit status", get_address("IRAM", OFF_IRAM_SUIT_STATUS), 1, updateSuitStatus, 60)
	bosses_status = ScriptHost:AddMemoryWatch("Boss status", get_address("IRAM", OFF_IRAM_BOSS_STATUS), 2, updateBossStatus, 60)
	lock_level = ScriptHost:AddMemoryWatch("Lock level", get_address("IRAM", OFF_IRAM_LOCK_LEVEL), 1, updateLockLevel, 60)
end

-- Clear memory watches
function clearMemoryWatches()
	ScriptHost:RemoveMemoryWatch(lock_level)
	ScriptHost:RemoveMemoryWatch(bosses_status)
	ScriptHost:RemoveMemoryWatch(suit_status)
	ScriptHost:RemoveMemoryWatch(beam_status)
	ScriptHost:RemoveMemoryWatch(max_etanks)
	ScriptHost:RemoveMemoryWatch(max_powerbombs)
	ScriptHost:RemoveMemoryWatch(max_missiles)
	ScriptHost:RemoveMemoryWatch(missile_bomb_status)
end